File: web/src/api/marketContext.ts
import { apiFetch } from "./client";
import type { FreshnessTier, ProviderName } from "./creditUniverse";

/**
 * One Market Context observation (backend/app/schemas/market_context.py).
 *
 * `value` arrives as a string, not a number — same Decimal-as-JSON-string
 * convention as `CreditUniverseRow` (see that file's docstring).
 */
export interface MarketContextObservation {
  series_id: string;
  title: string;
  value: string;
  units: string;
  as_of_date: string;
  freshness: FreshnessTier;
  provider: ProviderName;
}

/**
 * `null` for a field whenever that FRED series hasn't been synced yet —
 * never a fabricated placeholder value.
 */
export interface MarketContext {
  sofr: MarketContextObservation | null;
  high_yield_oas: MarketContextObservation | null;
}

export async function fetchMarketContext(): Promise<MarketContext> {
  return apiFetch<MarketContext>("/api/market-context");
}

File: backend/app/schemas/market_context.py
"""Response schema for the Market Context API.

Answers "what macroeconomic environment surrounds this credit?" (PLAN.md
section 18 step 5) directly and honestly: each field is a real FRED
observation shown as-is, with its own provenance/as-of date/freshness. No
blended or derived macro score — that's an explicit non-goal for this
milestone (see `services/market_context_service.py`).
"""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.core.freshness import FreshnessTier
from app.core.types import ProviderName


class MarketContextObservation(BaseModel):
    model_config = ConfigDict(frozen=True)

    series_id: str
    title: str
    value: Decimal
    units: str
    as_of_date: date
    freshness: FreshnessTier
    provider: ProviderName


class MarketContext(BaseModel):
    """`None` for a field whenever that series hasn't been synced yet —
    never a fabricated placeholder value."""

    model_config = ConfigDict(frozen=True)

    sofr: MarketContextObservation | None
    high_yield_oas: MarketContextObservation | None
